Update docs for building the code and contributing (#24703)
dotnet/efcore: .github/CONTRIBUTING.md

# How to contribute

We welcome community pull requests for bug fixes, enhancements, and documentation. Code and API documentation are in this repo. Conceptual documentation is in the [EntityFramework.Docs](https://github.com/dotnet/EntityFramework.Docs) repo.

> For code changes, you will need to [fork and clone the repo and build the code locally](./docs/getting-and-building-the-code.md).

Reviewer: Is the leading `./` needed? (here and below)

Reviewer 2: This is invalid URL

Author: I copy-pasted this from other places in the markdown. It seems to work. What is the recommended pattern?

## Choosing an issue

All contributions should address an [open issue](https://github.com/dotnet/efcore/issues) in the [dotnet/efcore](https://github.com/dotnet/efcore) repo.

> ### Bugs verses enhancements

Reviewer: ```suggestion
### Bugs versus enhancements
```

Author: This one was just for you, @roji!

Issues are typically labeled with [type-enhancement](https://github.com/dotnet/efcore/issues?q=is%3Aopen+is%3Aissue+label%3Atype-enhancement) or [type-bug](https://github.com/dotnet/efcore/issues?q=is%3Aopen+is%3Aissue+label%3Atype-bug).

* Bugs are places where EF Core is doing something that it was not designed to.
* Enhancements are suggestions to improve EF Core by changing existing or adding new functionality.

Bugs are usually relatively small, self-contained changes. However, this does not necessarily make them easy to work on, since finding the root cause and fixing it without breaking any other functionality can be tricky.

> Enhancements can be anything from tiny changes to support a variation on an existing scenario, to massive cross-cutting features that will take many months to implement. The bigger the enchantment, the more important it is to communicate with the EF Team before working on a contribution.

Reviewer: Enchantment :heart: 

```suggestion
Enhancements can be anything from tiny changes to support a variation on an existing scenario, to massive cross-cutting features that will take many months to implement. The bigger the enhancement, the more important it is to communicate with the EF Team before working on a contribution.
```

> Enhancements can be anything from tiny changes to support a variation on an existing scenario, to massive cross-cutting features that will take many months to implement. The bigger the enchantment, the more important it is to communicate with the EF Team before working on a contribution.

Reviewer 2: ```suggestion
Enhancements can be anything from tiny changes to support a variation on an existing scenario, to massive cross-cutting features that will take many months to implement. The bigger the enhancement, the more important it is to communicate with the EF Team before working on a contribution.
```

Author: This was a mistake but now I think we should switch to calling them all enchantments!

Issues labeled with [needs-design](https://github.com/dotnet/efcore/issues?q=is%3Aopen+is%3Aissue+label%3Aneeds-design) indicate that the EF Core team has not solidified on an approach to tackle the issue. This typically makes the issue much more difficult for the community to implement.

### Create an issue

If there is no existing issue tracking the change you want to make, then [create one](https://github.com/dotnet/efcore/issues/new/choose)! PRs that don't get merged are often those that are created without any prior discussion with the team. An issue is the best place to have that discussion, ideally before the PR is submitted.

### Fixing typos

An issue is not required for simple non-code changes like fixing a typo in documentation. In fact, these changes can often be submitted as a PR directly from the browser, avoiding the need to fork and clone.

## Workflow

The typical workflow for contributing to EF Core is outlined below. This is not a set-in-stone process, but rather guidelines to help ensure a quality PR that we can merge efficiently.

> * [Set up your development environment](./docs/getting-and-building-the-code.md) so that you can build and test the code. Don't forget to [create a fork](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo) for your work.

Reviewer 2: bad link

* Make sure all tests are passing. (This is typically done by running `test` at a command prompt.)
* Choose an issue (see above), understand it, and **comment on the issue** indicating what you intend to do fix it. **This communication with the team is very important and often helps avoid throwing away lots of work caused by taking the wrong approach.**

> * Create and checkout a [branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-and-deleting-branches-within-your-repository) in your local clone. You will use this branch to prepare your PR.

Reviewer: ```suggestion
* Create and check out a [branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-and-deleting-branches-within-your-repository) in your local clone. You will use this branch to prepare your PR.
```

* Make appropriate code and test changes. Follow the patterns and code style that you see in the existing code. Make sure to add tests that fail without the change and then pass with the change.

> * Consider other scenarios where your change may have an impact and add more testing. It's difficult to add too many tests.

Reviewer: Possibly:
```suggestion
* Consider other scenarios where your change may have an impact and add more testing. We always prefer having too many tests to having not enough of them.
```

> * Consider other scenarios where your change may have an impact and add more testing. It's difficult to add too many tests.

Reviewer 2: > It's difficult to add too many tests.

What does this mean? 🤔

Author: Will re-word.

```C#
public class Foo
{
    public Foo(string a)
        : this(a, null)
    {
    }

    public Foo(string a, string b)
    {
        A = a;
        B = b;
    }

    public string A { get; }
    public string B { get; }
}
```

This is now neither a source nor a binary breaking change.

### Behavioral breaking changes

Behavioral breaking changes happen when the behavior of calling some API changes without the API itself changing. For example, if calling `DbContext.Add` previously resulted in an entity in the `Added` state, and now results in it being in the `Modified` state, then that's a behavioral break. Existing applications will not be expecting a `Modified` entity and may therefore not behave correctly.

Behavioral breaking changes are hard to identify and are one of the reasons testing is so important. Running all the existing tests can help find behavioral breaks. Likewise, implementing good tests with your change helps ensure any future breaks will be discovered.

## Code-of-conduct

> Please remember to abide by our [code-of-conduct](.github/CODE_OF_CONDUCT.md) while working on contributions.

Reviewer 2: bad link
